contracts/utils/EnumerableSet.sol

```diff
@@ -0,0 +1,77 @@
+pragma solidity ^0.5.0;
+
+
+/**
+ * @title EnumerableSet
+ * @dev Data structure - https://en.wikipedia.org/wiki/Set_(abstract_data_type)
+ *
+ * An EnumerableSet.AddressSet is a data structure containing a number of unique addresses.
+ *
+ *  - It is possible to add and remove addresses in O(1).
+ *  - It is also possible to query if the AddressSet contains an address in O(1).
+ *  - It is possible to retrieve an array with all the addresses in the AddressSet using enumerate.
+ *    This operation is O(N) where N is the number of addresses in the AddressSet. The order in
+ *    which the addresses are retrieved is not guaranteed.
+ *
+ * Initialization of a set must include an empty array:
+ * `EnumerableSet.AddressSet set = EnumerableSet.AddressSet({values: new address[](0)});`
+ *
+ * @author Alberto Cuesta Cañada
+ */
+library EnumerableSet {
+
+    struct AddressSet {
+        // Position of the value in the `values` array, plus 1 because index 0
+        // means a value is not in the set.
+        mapping (address => uint256) index;
+        address[] values;
+    }
+
+    /**
+     * @dev Add a value. O(1).
+     */
+    function add(AddressSet storage set, address value)
+        internal
+    {
+        require(!contains(set, value), "EnumerableSet: value already in set");
```

Comment:
With flexibility in mind, and considering it is not possible to recover from a `revert` in an internal function, I think it'd be best to leave this up to the user, and return `false` when the call was a no-op and `true` when the data was indeed added/removed. Sorry to only reply after you've already made the change!

Reply:
Not a big issue, just a little extra code. You take the `revert` approach on `ERC20`, though, and everyone seems happy.

Reply:
I think there's a difference between those cases: we `revert` in ERC20 because the operation (transfer or approval) was _not_ performed, and we don't want to rely on the return value for errors. If somewhere where to send a transaction that calls `transfer` directly, they wouldn't even have access to the return value.

In this case, the function returning false doesn't mean that the value wasn't added to the set, but rather that it was already there. In both scenarios the outcome is the same: the value is in the set. 

Additionally, we're writing an `internal` library, to be used by other contracts. I'd rather those contracts decide how to handle these cases.

Reply:
You are right, there is quite some difference. Thanks for the explanation.